File: packages/vscode-extension/src/diff-apply-guard.js
/**
 * Pure safety guards for applying a reviewed diff back to disk. No `vscode`:
 * unit-testable, injected into the facade like diff-hunks / multi-diff.
 *
 *  - checkApplySafety: optimistic-concurrency check. The reviewer decided
 *    against the baseline the agent proposed from (`originalText`); if the
 *    on-disk file changed during the review, a blind whole-file write would
 *    silently destroy those concurrent edits. When no baseline was provided
 *    (older callers diff against the live file) the answer is always "safe"
 *    so the legacy path stays byte-identical.
 *
 *  - looksBinary: NUL-byte sniff (git's heuristic) so the text diff pipeline
 *    never round-trips a binary file through UTF-8 — that corrupts the bytes.
 *    Multi-byte UTF-8 text (中文 etc.) never contains a NUL and is not
 *    misdetected.
 *
 * Twin: packages/jetbrains-plugin/src/main/java/com/chainlesschain/ide/
 * DiffApplyGuard.java — keep the semantics in sync.
 */

/** Reason string for a drifted-disk verdict (shared with the JB twin). */
const REASON_DISK_DRIFTED = "disk-drifted";

/** Result `reason` for a binary short-circuit (shared with the JB twin). */
const REASON_BINARY_SKIPPED = "binary file, skipped";

/** Combined baseline + proposal budget for one native IDE diff. */
const MAX_REVIEW_FILE_BYTES = 2 * 1024 * 1024;

/** Result `reason` when a text diff is too large for a safe native review. */
const REASON_LARGE_FILE_SKIPPED = "file too large for IDE diff review";

/**
 * @param {object} args
 * @param {string} [args.baselineText] the baseline the review was proposed
 *   against (openDiff's `originalText`); non-string → no baseline → safe.
 * @param {string} [args.currentDiskText] the file's current on-disk content;
 *   non-string (unreadable / missing file) → nothing to clobber → safe.
 * @returns {{safe: true} | {safe: false, reason: string}}
 */
function checkApplySafety({ baselineText, currentDiskText } = {}) {
  if (typeof baselineText !== "string" || typeof currentDiskText !== "string") {
    return { safe: true };
  }
  return baselineText === currentDiskText
    ? { safe: true }
    : { safe: false, reason: REASON_DISK_DRIFTED };
}

/**
 * NUL-byte binary sniff. Accepts a string, Buffer or Uint8Array; anything
 * else (null/undefined included) is "not binary".
 * @returns {boolean}
 */
function looksBinary(input) {
  if (typeof input === "string") {
    return input.includes("\u0000");
  }
  if (input instanceof Uint8Array) {
    // Buffer extends Uint8Array; indexOf(0) is a native byte scan.
    return input.indexOf(0) !== -1;
  }
  return false;
}

function byteLength(input) {
  if (typeof input === "string") return Buffer.byteLength(input, "utf8");
  if (input instanceof Uint8Array) return input.byteLength;
  return 0;
}

/**
 * Classify one baseline/proposal pair before opening an IDE text diff.
 * Binary input is rejected first; otherwise the combined UTF-8 payload is
 * bounded so a host cannot be forced to materialize arbitrarily large editor
 * documents. When no explicit baseline is supplied, `currentBytes` is a
 * bounded binary probe and `currentSizeBytes` is the full on-disk size.
 */
function checkReviewPayload({
  modifiedText,
  originalText,
  currentBytes,
  currentSizeBytes,
  maxBytes = MAX_REVIEW_FILE_BYTES,
} = {}) {
  const baseline =
    typeof originalText === "string" ? originalText : currentBytes;
  const limit =
    Number.isSafeInteger(maxBytes) && maxBytes > 0
      ? maxBytes
      : MAX_REVIEW_FILE_BYTES;
  const baselineBytes =
    typeof originalText === "string"
      ? byteLength(originalText)
      : Number.isSafeInteger(currentSizeBytes) && currentSizeBytes >= 0
        ? currentSizeBytes
        : byteLength(baseline);
  const bytes = byteLength(modifiedText) + baselineBytes;
  if (looksBinary(modifiedText) || looksBinary(baseline)) {
    return {
      reviewable: false,
      kind: "binary",
      reason: REASON_BINARY_SKIPPED,
      bytes,
      limitBytes: limit,
    };
  }
  if (bytes > limit) {
    return {
      reviewable: false,
      kind: "large-file",
      reason: REASON_LARGE_FILE_SKIPPED,
      bytes,
      limitBytes: limit,
    };
  }
  return { reviewable: true, bytes, limitBytes: limit };
}

module.exports = {
  checkApplySafety,
  checkReviewPayload,
  looksBinary,
  MAX_REVIEW_FILE_BYTES,
  REASON_DISK_DRIFTED,
  REASON_BINARY_SKIPPED,
  REASON_LARGE_FILE_SKIPPED,
};

File: packages/jetbrains-plugin/src/main/java/com/chainlesschain/ide/DiffApplyGuard.java
package com.chainlesschain.ide;

import java.nio.charset.StandardCharsets;

/**
 * Pure safety guards for applying a reviewed diff back to disk. SDK-free so
 * plain JUnit covers it; the IntelliJ glue lives in IntellijEditorFacade.
 *
 * <ul>
 *   <li>{@link #safeToApply}: optimistic-concurrency check. The reviewer
 *       decided against the baseline the agent proposed from
 *       ({@code originalText}); if the file changed during the review, a
 *       blind whole-file write would silently destroy those concurrent
 *       edits. No baseline (legacy callers diff against the live file) or no
 *       readable current text (nothing to clobber) → always safe, keeping
 *       the old call paths byte-identical.</li>
 *   <li>{@link #looksBinary}: NUL sniff (git's heuristic) so the text diff
 *       pipeline never round-trips a binary file through UTF-8 — that
 *       corrupts the bytes. Multi-byte UTF-8 text (中文 etc.) contains no
 *       NUL and is never misdetected.</li>
 * </ul>
 *
 * Twin: packages/vscode-extension/src/diff-apply-guard.js — keep semantics
 * in sync.
 */
public final class DiffApplyGuard {

    /** Result {@code reason} when a drifted disk blocks the apply. */
    public static final String REASON_DISK_DRIFTED = "disk-drifted";

    /** Result {@code reason} for a binary short-circuit. */
    public static final String REASON_BINARY_SKIPPED = "binary file, skipped";

    /** Combined baseline + proposal budget for one native IDE diff. */
    public static final long MAX_REVIEW_FILE_BYTES = 2L * 1024L * 1024L;

    /** Result {@code reason} for an oversized text diff. */
    public static final String REASON_LARGE_FILE_SKIPPED =
            "file too large for IDE diff review";

    /** Bounded, content-free classification returned before an IDE diff opens. */
    public static final class ReviewPayload {
        public final boolean reviewable;
        public final String kind;
        public final String reason;
        public final long bytes;
        public final long limitBytes;

        ReviewPayload(
                boolean reviewable,
                String kind,
                String reason,
                long bytes,
                long limitBytes) {
            this.reviewable = reviewable;
            this.kind = kind;
            this.reason = reason;
            this.bytes = bytes;
            this.limitBytes = limitBytes;
        }
    }

    private DiffApplyGuard() {
    }

    /**
     * @param baselineText the baseline the review was proposed against
     *                     (openDiff's {@code originalText}); null → no
     *                     baseline → safe.
     * @param currentText  the file's current content; null (missing /
     *                     unreadable file) → nothing to clobber → safe.
     * @return whether a whole-file write may proceed without destroying
     *         concurrent edits.
     */
    public static boolean safeToApply(String baselineText, String currentText) {
        if (baselineText == null || currentText == null) return true;
        return baselineText.equals(currentText);
    }

    /** NUL-char sniff on text; null is "not binary". */
    public static boolean looksBinary(String text) {
        return text != null && text.indexOf('\0') >= 0;
    }

    /** NUL-byte sniff on raw bytes; null is "not binary". */
    public static boolean looksBinary(byte[] bytes) {
        if (bytes == null) return false;
        for (byte b : bytes) {
            if (b == 0) return true;
        }
        return false;
    }

    /**
     * Classify a baseline/proposal pair before creating native editor
     * documents. {@code currentBytes} is used only when no explicit baseline
     * was supplied.
     */
    public static ReviewPayload checkReviewPayload(
            String modifiedText,
            String originalText,
            byte[] currentBytes,
            long maxBytes) {
        return checkReviewPayload(
                modifiedText,
                originalText,
                currentBytes,
                currentBytes == null ? null : Long.valueOf(currentBytes.length),
                maxBytes);
    }

    public static ReviewPayload checkReviewPayload(
            String modifiedText,
            String originalText,
            byte[] currentBytes,
            Long currentSizeBytes,
            long maxBytes) {
        long limit = maxBytes > 0 ? maxBytes : MAX_REVIEW_FILE_BYTES;
        boolean explicitBaseline = originalText != null;
        long bytes = utf8Length(modifiedText)
                + (explicitBaseline
                        ? utf8Length(originalText)
                        : currentSizeBytes != null && currentSizeBytes >= 0
                                ? currentSizeBytes
                                : byteLength(currentBytes));
        boolean binary = looksBinary(modifiedText)
                || (explicitBaseline ? looksBinary(originalText) : looksBinary(currentBytes));
        if (binary) {
            return new ReviewPayload(
                    false, "binary", REASON_BINARY_SKIPPED, bytes, limit);
        }
        if (bytes > limit) {
            return new ReviewPayload(
                    false, "large-file", REASON_LARGE_FILE_SKIPPED, bytes, limit);
        }
        return new ReviewPayload(true, null, null, bytes, limit);
    }

    public static ReviewPayload checkReviewPayload(
            String modifiedText,
            String originalText,
            byte[] currentBytes) {
        return checkReviewPayload(
                modifiedText, originalText, currentBytes, MAX_REVIEW_FILE_BYTES);
    }

    private static long utf8Length(String text) {
        return text == null ? 0L : text.getBytes(StandardCharsets.UTF_8).length;
    }

    private static long byteLength(byte[] bytes) {
        return bytes == null ? 0L : bytes.length;
    }
}
